lib/compiler/src/beam_doc.erl

```diff
@@ -0,0 +1,241 @@
+documentation format.
+".
+
+-record(docs, {cwd                 :: unicode:chardata(),             % Cwd
+               module_name = undefined  :: unicode:chardata() | undefined,
+               doc    = undefined  :: unicode:chardata(), % Function/type/callback local doc
+               meta   = maps:new() :: map()}).      % Function/type/callback local meta
+
+-doc "
+Internal documentation type for generating EEP-48 documentation format.
+".
+-type internal_docs() :: #docs{}.
+
+
+-doc "
+Transforms an Erlang abstract syntax form into EEP-48 documentation format.
+".
+-spec main(term(), term()) -> term().
+main(Dirname, AST) ->
+    try
+        {ModuleDocAnno, ModuleDoc} = extract_moduledoc(AST),
+        DocFormat = extract_docformat(AST),
+        Docs = extract_documentation(AST, new_state(Dirname)),
+        #docs_v1{
+           format = DocFormat,
+           anno = ModuleDocAnno,
+           module_doc = #{ <<"en">> => ModuleDoc },
+           docs = Docs }
+    catch E:R:ST ->
+            erlang:raise(E, R, ST)
+    end.
+
+-spec extract_docformat(AST :: [tuple()]) -> DocFormat :: binary().
+extract_docformat(AST) ->
+    case lists:keysearch(docformat, 3, AST) of
+        false -> <<"text/markdown">>;
+        {value, {attribute, _DocAnno, docformat, DocFormat}} ->
+            unicode:characters_to_binary(DocFormat)
+    end.
+
+-spec extract_moduledoc(AST :: [tuple()]) -> ModuleDoc :: {erl_anno:anno(), binary()}.
+extract_moduledoc(AST) ->
+    {value, {attribute, ModuleDocAnno, moduledoc, ModuleDoc}} = lists:keysearch(moduledoc, 3, AST),
+    {ModuleDocAnno, unicode:characters_to_binary(string:trim(ModuleDoc))}.
+
+-spec new_state(Dirname :: unicode:chardata()) -> internal_docs().
+new_state(Dirname) ->
+    reset_state(#docs{cwd = Dirname}).
+
+-spec reset_state(State :: internal_docs()) -> internal_docs().
+reset_state(State) ->
+    State#docs{doc = undefined, meta = maps:new()}.
+
+-spec update_meta(State :: internal_docs(), Meta :: map()) -> internal_docs().
+update_meta(#docs{meta = Meta0}=State, Meta1) ->
+    State#docs{meta = maps:merge(Meta0, Meta1)}.
+
+-spec update_doc(State :: internal_docs(), Doc :: unicode:chardata()) -> internal_docs().
+update_doc(#docs{}=State, Doc) ->
+    State#docs{doc = string:trim(Doc)}.
+
+-spec update_module(State :: internal_docs(), ModuleName :: unicode:chardata()) -> internal_docs().
+update_module(#docs{}=State, ModuleName) ->
+    State#docs{module_name = ModuleName}.
+
+
+-doc "
+Creates EEP-48 documentation format from an Erlang abstract form.
+".
+extract_documentation([{attribute, _Anno, file, {ModuleName, _A}} | T], State) ->
+    extract_documentation(T, update_module(State, ModuleName));
+extract_documentation([{attribute, _Anno, doc, Meta0} | T], State) when is_map(Meta0) ->
+    extract_documentation(T, update_meta(State, Meta0));
+extract_documentation([{attribute, _Anno, doc, {file, Path}}|T], #docs{cwd = Cwd}=State) ->
```

Comment:
Is this clause needed?

Reply:
Thanks, this was solved already by the epp.